### DISCUSSION TITLE
Applying metrics

### DISCUSSION THREAD

**Discussion by Meddebma** [Project-MONAI/MONAI#1696] [Q&A] [answered]
Hi, 

I was evaluating the unet for spleen segmentation with different metrics, but I got en Error trying to compute the ROC:

here is the code: 
with torch.no_grad():
        metric_sum = 0.0
        metric_count = 0
        for val_data in val_loader:
            # define sliding window size and batch size for windows inference
            val_images, val_labels = val_data["image"].to(device), val_data["label"].to(device)
            roi_size = (96, 96, 96)
            sw_batch_size = 4
            val_outputs = sliding_window_inference(val_data["image"].to(device), roi_size, sw_batch_size, model)
            val_outputs = post_trans(val_outputs)
            value = compute_meandice(y_pred=val_outputs, y=val_labels, include_background=False)
            roc = compute_roc_auc(y_pred=val_outputs, y=val_labels)
            conf = get_confusion_matrix(y_pred=val_outputs, y=val_labels, include_background=False)
            hausdorff = compute_hausdorff_distance(y_pred=val_outputs, y=val_labels, include_background=False)
            metric_count += len(value)
            metric_sum += value.item() * len(value)
            #print(f"val_loss:", loss)
            print(f"val_dice:", value)
            print(f"Confusion Matrix:", conf)
            print(f"Hausdorff:",hausdorff)
            print(f"roc_auc:", roc)

and here is the Error:
---------------------------------------------------------------------------

ValueError                                Traceback (most recent call last)

<ipython-input-17-f4a81bc34b4e> in <module>()
     30             y=val_labels,
     31             include_background=False)
---> 32             roc = compute_roc_auc(y_pred=val_outputs, y=val_labels)
     33             conf = get_confusion_matrix(y_pred=val_outputs, y=val_labels, include_background=False)
     34             hausdorff = compute_hausdorff_distance(y_pred=val_outputs, y=val_labels, include_background=False)

/usr/local/lib/python3.7/dist-packages/monai/metrics/rocauc.py in compute_roc_auc(y_pred, y, to_onehot_y, softmax, other_act, average)
    100     y_ndim = y.ndimension()
    101     if y_pred_ndim not in (1, 2):
--> 102         raise ValueError("Predictions should be of shape (batch_size, n_classes) or (batch_size, ).")
    103     if y_ndim not in (1, 2):
    104         raise ValueError("Targets should be of shape (batch_size, n_classes) or (batch_size, ).")

ValueError: Predictions should be of shape (batch_size, n_classes) or (batch_size, ).

How should I fix this?

Best, Aymen

**Comment by rijobro** [comment, marked as answer]
As the error indicates, the `y_pred` of `compute_roc_auc` is expected to be a 1- or 2D array with the predicted classes. In your case, however, you are passing your segmented image.

Some metrics make sense for classification but don't make sense for segmentation. ROC/AUC and confusion matrices are examples of this (e.g., see [this blog post](https://towardsdatascience.com/understanding-the-roc-and-auc-curves-a05b68550b69)).

**Reply by Meddebma** [reply]
Thanks for the reading, this makes totally sense!
